### PR TITLE
Vm clone

### PR DESCRIPTION
**Pull request by ahmedelsayed-93**
fixes https://github.com/0-complexity/G8_testing/issues/242

### CODE REVIEW

File: functional_testing/Openvcloud/ovc_master_hosted/OVC/c_advanced/machine_tests.py [hunk 1]
```diff
@@ -366,16 +366,40 @@ def test011_restart_vm_after_migration(self):
...
     def test012_check_cloned_vm(self):
-        """ OVC-000
+        """ OVC-029
         *Test case for checking cloned VM ip, portforwards and credentials*
-
         **Test Scenario:**
-
-        #. Create (VM1).
+        #. Create (VM1), should succeed.
+        #. Create portforward to ssh port for (VM1).
```
Comment (john-kheir) [inline functional_testing/Openvcloud/ovc_master_hosted/OVC/c_advanced/machine_tests.py]: Please take a snapshot for vm1 then write a file on the machine and after the clone make sure this file is there and that u can't rollback using the parent vm snapshot

File: functional_testing/Openvcloud/ovc_master_hosted/OVC/c_advanced/machine_tests.py [hunk 2]
```diff
@@ -366,16 +366,40 @@ def test011_restart_vm_after_migration(self):
...
+        #. Stop (VM1), should succeed.
         #. Clone VM1 as (VM2_C), should succeed.
         #. Make sure VM2_C got a new ip.
-        #. Check that VM2 got different credentials from VM1.
         #. Make sure no portforwards have been created.
         """
+        self.lg('Create (VM1), should succeed')
+        machineId = self.cloudapi_create_machine(self.cloudspace_id)
+        self.wait_for_machine_to_get_ip(machineId)
```
Comment (john-kheir) [inline functional_testing/Openvcloud/ovc_master_hosted/OVC/c_advanced/machine_tests.py]: check on the ip before proceeding .. in this case u may return False in wait_for_machine_to_get_ip

File: functional_testing/Openvcloud/ovc_master_hosted/OVC/c_advanced/machine_tests.py (hunk 1 above)
Comment (ahmedelsayed-93) [reply functional_testing/Openvcloud/ovc_master_hosted/OVC/c_advanced/machine_tests.py]: done

File: functional_testing/Openvcloud/ovc_master_hosted/OVC/c_advanced/machine_tests.py (hunk 2 above)
Comment (ahmedelsayed-93) [reply functional_testing/Openvcloud/ovc_master_hosted/OVC/c_advanced/machine_tests.py]: done

**Comment by ahmedelsayed-93** [comment]
```
root@be-g8-4-ctrl01:/tmp/G8_testing/functional_testing/Openvcloud# nosetests-2.7 -s -v ovc_master_hosted/OVC/c_advanced/machine_tests.py:MachineTests.test012_check_cloned_vm --tc-file config.ini
OVC-029 ... HTTP Error 409: Conflict
ok

----------------------------------------------------------------------
Ran 1 test in 147.736s

OK
```

File: functional_testing/Openvcloud/ovc_master_hosted/OVC/c_advanced/machine_tests.py [hunk 3]
```diff
@@ -366,16 +366,64 @@ def test011_restart_vm_after_migration(self):
...
+        #. Take a snapshot (SS0) for (VM1).
+        #. Write file (F1) on (VM1).
+        #. Stop (VM1), should succeed.
         #. Clone VM1 as (VM2_C), should succeed.
+        #. Start (VM1), should succeed
         #. Make sure VM2_C got a new ip.
-        #. Check that VM2 got different credentials from VM1.
         #. Make sure no portforwards have been created.
+        #. Check that file (F1) exists.
+        #. Rollback (VM2_C) to snapshot (SS1), shoud fail.
```
Comment (john-kheir) [inline functional_testing/Openvcloud/ovc_master_hosted/OVC/c_advanced/machine_tests.py]: shoud -- > should

File: functional_testing/Openvcloud/ovc_master_hosted/OVC/c_advanced/machine_tests.py [hunk 4]
```diff
@@ -366,16 +366,64 @@ def test011_restart_vm_after_migration(self):
...
+
+        self.lg('Create portforward to ssh port for (VM1)')
+        self.get_vm_ssh_publicport(machineId, wait_vm_ip=False)
+
+        self.lg('Take a snapshot (SS0) for (VM1)')
+        snapshotId = self.api.cloudapi.machines.snapshot(machineId=machineId, name='test-snapshot')
+        snapshotEpoch = self.api.cloudapi.machines.listSnapshots(machineId=machineId)[0]['epoch']
+
+        self.lg('Write file to (VM1)')
+        machine_1_connection = self.get_vm_connection(machineId, wait_vm_ip=False)
```
Comment (john-kheir) [inline functional_testing/Openvcloud/ovc_master_hosted/OVC/c_advanced/machine_tests.py]: touch the file is enough

File: functional_testing/Openvcloud/utils/utils.py [hunk 5]
```diff
@@ -291,6 +291,18 @@ def wait_for_status(self, status, func, timeout=300, **kwargs):
...
+    def wait_for_machine_to_get_ip(self, machineId, timeout=300):
+        for _ in range(timeout):
+            machine_info = self.api.cloudapi.machines.get(machineId=machineId)
+            ip_address = machine_info['interfaces'][0]['ipAddress']
+            if ip_address != 'Undefined':
+                return ip_address
+            else:
+                time.sleep(1)
+        else:
+            return 'Undefined'
```
Comment (john-kheir) [inline functional_testing/Openvcloud/utils/utils.py]: return None or False instead of Undefined

**Pull request merged**